src/box/applier.cc

```diff
@@ -207,10 +207,8 @@ applier_thread_writer_f(va_list ap)
 		try {
 			applier->thread.has_acks_to_send = false;
 			struct xrow_header xrow;
-			uint64_t vclock_sync = applier->thread.last_vclock_sync;
-			struct vclock *ack_vclock = &applier->thread.ack_vclock;
-			xrow_encode_heartbeat_xc(&xrow, ack_vclock,
-						 vclock_sync);
+			xrow_encode_upstream_ack_xc(&xrow,
+						    &applier->thread.next_ack);
 			xrow.tm = applier->thread.txn_last_tm;
```

Comment:
I don't remember a discussion. Maybe you talked to Vova. My opinion here is that adding `tm` and `replica_id` to request wouldn't be better, because:
- It would be duplication of the members. You would have `tm` and `replica_id` both in `xrow_header` and in `..._request` structs.
- Duplication might lead to slightly bigger parent objects. For example, `relay.last_recv_ack` would store `tm` field even though it is not needed here. The timestamp is decoded and used in `relay_reader_f()` right away, doesn't need a long storage.

Doesn't mean it can't be done, but it would simply produce other questions. Can do it if you want anyway.

Reply:
I'm fine with the current approach, thanks for the answer.